### PR TITLE
AJ-1263 Fix `IllegalStateException` on startup while polling SAM status.

### PR DESCRIPTION
**Pull request by jladieu**
AJ-1263 Fix `IllegalStateException` on startup while polling SAM status.

When starting WDS locally, the startup process attempts to poll SAM status outside the context of a user request.  Spring throws an error when `RequestContextHolder.currentRequestAttributes` is called outside the context of a web request, which caused the below crash indicating SAM was down.

This fix adds a check to only use `currentRequestAttributes` in the context of a request.  Otherwise the reqeust will be treated as unauthenticated via an access token.

```
2023-08-10 11:32:16.535  INFO [] [restartedMain] o.d.w.logging.AvailabilityListener       : ReadinessState: ACCEPTING_TRAFFIC
2023-08-10 11:32:16.946 DEBUG [] [RMI TCP Connection(2)-10.1.7.98] o.d.w.sam.HttpSamClientSupport           : Sending getSystemStatus request to Sam ...
2023-08-10 11:32:16.951 ERROR [] [RMI TCP Connection(2)-10.1.7.98] o.d.w.sam.HttpSamClientSupport           : getSystemStatus Sam request resulted in No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request.

java.lang.IllegalStateException: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request.
	at org.springframework.web.context.request.RequestContextHolder.currentRequestAttributes(RequestContextHolder.java:131)
	at org.databiosphere.workspacedataservice.sam.HttpSamClientFactory.getApiClient(HttpSamClientFactory.java:50)
	at org.databiosphere.workspacedataservice.sam.HttpSamClientFactory.getStatusApi(HttpSamClientFactory.java:85)
	at org.databiosphere.workspacedataservice.sam.HttpSamDao.lambda$getSystemStatus$2(HttpSamDao.java:111)
	at org.databiosphere.workspacedataservice.sam.HttpSamClientSupport.withRetryAndErrorHandling(HttpSamClientSupport.java:57)
	at org.databiosphere.workspacedataservice.sam.HttpSamClientSupport$$FastClassBySpringCGLIB$$92d51fe7.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:97)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:133)
	at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:159)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at org.databiosphere.workspacedataservice.sam.HttpSamClientSupport$$EnhancerBySpringCGLIB$$5e50f93b.withRetryAndErrorHandling(<generated>)
	at org.databiosphere.workspacedataservice.sam.HttpSamDao.getSystemStatus(HttpSamDao.java:112)
	at org.databiosphere.workspacedataservice.sam.HttpSamDao.getSystemStatusOk(HttpSamDao.java:107)
	at org.databiosphere.workspacedataservice.sam.HttpSamDao$$FastClassBySpringCGLIB$$dca659b5.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.cache.interceptor.CacheInterceptor.lambda$invoke$0(CacheInterceptor.java:54)
	at org.springframework.cache.interceptor.CacheAspectSupport.invokeOperation(CacheAspectSupport.java:366)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:421)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:345)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:64)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at org.databiosphere.workspacedataservice.sam.HttpSamDao$$EnhancerBySpringCGLIB$$f9be12c4.getSystemStatusOk(<generated>)
	at org.databiosphere.workspacedataservice.sam.PermissionsStatusService.doHealthCheck(PermissionsStatusService.java:29)
	at org.springframework.boot.actuate.health.AbstractHealthIndicator.health(AbstractHealthIndicator.java:82)
	at org.springframework.boot.actuate.health.HealthIndicator.getHealth(HealthIndicator.java:37)
	at org.springframework.boot.actuate.health.HealthEndpoint.getHealth(HealthEndpoint.java:94)
	at org.springframework.boot.actuate.health.HealthEndpoint.getHealth(HealthEndpoint.java:41)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getLoggedHealth(HealthEndpointSupport.java:172)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getContribution(HealthEndpointSupport.java:145)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getAggregateContribution(HealthEndpointSupport.java:156)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getContribution(HealthEndpointSupport.java:141)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getHealth(HealthEndpointSupport.java:110)
	at org.springframework.boot.actuate.health.HealthEndpointSupport.getHealth(HealthEndpointSupport.java:81)
	at org.springframework.boot.actuate.health.HealthEndpoint.health(HealthEndpoint.java:88)
	at org.springframework.boot.actuate.health.HealthEndpoint.health(HealthEndpoint.java:78)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:282)
	at org.springframework.boot.actuate.endpoint.invoke.reflect.ReflectiveOperationInvoker.invoke(ReflectiveOperationInvoker.java:74)
	at org.springframework.boot.actuate.endpoint.annotation.AbstractDiscoveredOperation.invoke(AbstractDiscoveredOperation.java:60)
	at org.springframework.boot.actuate.endpoint.jmx.EndpointMBean.invoke(EndpointMBean.java:124)
	at org.springframework.boot.actuate.endpoint.jmx.EndpointMBean.invoke(EndpointMBean.java:97)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:814)
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:802)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1472)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1310)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1405)
	at java.management.rmi/javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:829)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

2023-08-10 11:32:16.951  WARN [] [RMI TCP Connection(2)-10.1.7.98] o.d.w.retry.RetryLoggingListener         : Retryable method public <T> T org.databiosphere.workspacedataservice.sam.HttpSamClientSupport.withRetryAndErrorHandling(org.databiosphere.workspacedataservice.sam.HttpSamClientSupport$SamFunction<T>,java.lang.String) throws org.databiosphere.workspacedataservice.service.model.exception.SamException,org.databiosphere.workspacedataservice.service.model.exception.AuthenticationException,org.databiosphere.workspacedataservice.service.model.exception.AuthorizationException threw 1th exception org.databiosphere.workspacedataservice.service.model.exception.SamException: 500 INTERNAL_SERVER_ERROR "Error from Sam: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request."
2023-08-10 11:32:16.951  WARN [] [RMI TCP Connection(2)-10.1.7.98] o.d.w.sam.PermissionsStatusService       : SAM is currently signaled as DOWN.
```

### CODE REVIEW

**Comment by jladieu** [comment]
I don't have a good sense of how risky/safe this change is, so would love input particularly on whether this is _really not a good idea_

File: service/src/test/java/org/databiosphere/workspacedataservice/pact/SamPactTest.java [hunk 1]
```diff
@@ -29,14 +27,6 @@ class SamPactTest {
 
     static final String dummyResourceId = "92276398-fbe4-414a-9304-e7dcf18ac80e";
 
-    @BeforeAll
```
Comment (jladieu) [inline service/src/test/java/org/databiosphere/workspacedataservice/pact/SamPactTest.java:32]: This isn't necessary anymore after the other changes :)

**Comment by calypsomatic** [comment]
The code *looks* good but when I run it all calls are deemed unauthorized.  Trying to figure out what's up.

File: service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java [hunk 2]
```diff
@@ -37,40 +38,45 @@ public HttpSamClientFactory(String samUrl) {
...
+    }
+
     /**
      * Get a ResourcesApi Sam client, initialized with the url to Sam and the current user's
      * access token, if any
      * @return the usable Sam client
      */
     public ResourcesApi getResourcesApi(String token) {
-        ApiClient apiClient = getApiClient(token);
+        ApiClient apiClient = getApiClient();
```
Comment (calypsomatic) [inline service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: If the token is null, then we need to use `getCurrentUserBearerToken` - which may or may not be null, but you need to check.  Same with `getUsersApi` below.

Comment (jladieu) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: Thanks for pointing this out...is this related to what you observed in your other comment where all the calls are deemed unauthorized, or is this an orthogonal comment?

Comment (calypsomatic) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: I believe it's the reason why all the calls are deemed unauthorized - they were given a null token instead of the bearer token.

Comment (jladieu) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: Got it...possibly too much of a detour, but I'd love to be able to get the compiler to help us protect against this and use `@NonNull` annotation to push the concern upstream (and thus make the guarantee that if this gets called, a token has already been determined).  Having the logic flip/flop based on the nullability of the input feels brittle.  If this doesn't sound unreasonable, I can scout around to see if I can adjust the callers to support this.

Comment (calypsomatic) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: I have no opinion on this, so go for it.  See if you can get the whole code base cleaned up by the time you finish setting up your dev environment while you're at it ;p

Comment (calypsomatic) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: We could also consider if we need to adjust any unit tests to catch when we should/shouldn't have a null bearer token.  I think right now they only test that null tokens result in rejects.

**Comment by davidangb** [comment]
I can't replicate the error when running WDS locally on `main` - @calypsomatic can you? I know that "No thread-bound request found" error has hit us in a few other places.

Assuming this PR moves forward, let's get it a better Jira ticket, even if we use AJ-1185 for it.

Related: AJ-1159 and AJ-1160; should we need to make these same changes for the Leo, WSM, TDR clients?

File: service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java (hunk 2 above)
Comment (jladieu) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:78]: Ok I'm seeing now that the presence or absence of a token param is baked in several layers up the callstack for permission checks.  I think this might be too much of an undertaking to include in this PR, but may be worth potentially revisiting in the future to norm on having one access pattern for permissions.

IMO, ideally we'd have either / or -- _always_ get and pass a token to permission checks vs. never have to worry about doing so.

Comment (jladieu) [reply service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java]: To fix this I basically moved the token back into the `getApiClient` call where it was originally, and  tried to at least make the use of `null` intention revealing throughout by using a new constant `USE_BEARER_TOKEN_IF_PRESENT`

**Comment by jladieu** [comment]
I added a new ticket explicitly for this bug: https://broadworkbench.atlassian.net/browse/AJ-1263?atlOrigin=eyJpIjoiYjA4YWFmZTYyZWUxNGEyMzg3MDc4NTJkMzU5MDA4NDgiLCJwIjoiaiJ9

While the branch name for this PR will still use the old ref, I've updated the comments & PR title to reflect the new ticket.

**Comment by yuliadub** [comment]
> I can't replicate the error when running WDS locally on `main` - @calypsomatic can you? I know that "No thread-bound request found" error has hit us in a few other places.
> 
> Assuming this PR moves forward, let's get it a better Jira ticket, even if we use AJ-1185 for it.
> 
> Related: AJ-1159 and AJ-1160; should we need to make these same changes for the Leo, WSM, TDR clients?



> I can't replicate the error when running WDS locally on `main` - @calypsomatic can you? I know that "No thread-bound request found" error has hit us in a few other places.
> 
> Assuming this PR moves forward, let's get it a better Jira ticket, even if we use AJ-1185 for it.
> 
> Related: AJ-1159 and AJ-1160; should we need to make these same changes for the Leo, WSM, TDR clients?

I do not see the error when running this on main locally. Weird!

I would say if we make changes to SAM client, we might as well do it for all the others ones :) (Leo, WSM, TDR) might as well rope 1159 into it too!

**Comment by jladieu** [comment]
Closing this PR until I find evidence of recurrence.  Some of the refactoring herein may still be worthwhile, but I'll do those as a separate PR that links back to this one if I find motivation to pursue them.

File: service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java [hunk 3]
```diff
@@ -45,28 +48,40 @@ private ApiClient getApiClient(String accessToken) {
...
+        Optional.ofNullable(accessToken)
+                .or(HttpSamClientFactory::getCurrentUserBearerToken)
+                .ifPresent(apiClient::setAccessToken);
+
         // return the client
         return apiClient;
     }
 
+    private static Optional<String> getCurrentUserBearerToken() {
+        if (Objects.isNull(RequestContextHolder.getRequestAttributes())) {
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:63]: can check for params if you don't want to deal with the `IllegalStateException`

File: service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java [hunk 4]
```diff
@@ -22,6 +23,8 @@
  */
 public class HttpSamClientFactory implements SamClientFactory {
 
+    public static final String USE_BEARER_TOKEN_IF_PRESENT = null;
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/sam/HttpSamClientFactory.java:26]: crosslink for context
